feat: Implement Carousel component

dzole0311 opened the pull request:
Three core usage examples are available on the Disasters theme page: https://deploy-preview-25--next-earth-gov.netlify.app/themes/disasters

- Interact with the carousel and verify that it works as expected
- Only verify the carousel functionality and behavior using the examples above. We'll assemble the final pages with the carousel once this core component is merged (e.g the topic card still needs to be adjusted style wise etc.)
- @faustoperez @esthersuh any feedback on the transitions? I couldn’t find the exact values for fade speed/duration from the prototypes, so I tried to approximate them as closely as possible

dzole0311 commented on package.json:
```diff
@@ -23,10 +23,12 @@
     "@mdx-js/react": "^3.1.0",
     "@next/mdx": "^15.3.0",
-    "@teamimpact/veda-ui": "^6.1.1",
+    "@teamimpact/veda-ui": "^6.5.2-dev-test",
```
To be removed before merging, we did not export the Topic card from the veda-ui so I had to publish a dev package for testing.

dzole0311 commented on app/content/themes/disasters.mdx:
```diff
@@ -7,6 +7,152 @@ image: "/images/themes/disasters/disasters-banner.png"
 <Section>
   <Title>INFO</Title>
 
+<Section full>
```
Best to move these to a blank page with examples that's also not bundled and published to prod. Since we don't have that yet, I've added a few showcases for the carousel on one of the Theme pages for now - to be removed before merging.

dzole0311 commented on app/components/common/ClientOnly.tsx:
```diff
@@ -0,0 +1,27 @@
+ * This keeps the rest of the server-rendered content intact (e.g. for SEO or when JS is disabled),
+ * and prevents runtime errors when a client-only component is used in that context (e.g. as a carousel slide).
+ */
+export default function ClientOnly({
```
I ran into an issue when using components from veda-ui in a purely server-rendered page (e.g. Themes page). Even when importing simple tags from veda-ui that don’t use browser APIs, Next.js throws a `ReferenceError` related to the window object missing. Maybe the issue comes from how veda-ui is bundled and part of the client-only logic there leaks into the server render phase in Next.js? 

I tried guarding some of the remaining window references inside veda-ui with `typeof window !== 'undefined'` but that didn’t resolve the issue somehow. So I made this `ClientOnly` wrapper to create a client-only rendering boundary (like a “slot”) that defers the hydration until after the page has rendered on the server.

faustoperez reviewed (changes requested):
Looking great Gjore! Just two minor comments:

- [ ] In the card carousel I think the steps are misaligned, it seems like there's an extra step in the logic that doesn't match the card position

<img width="1494" alt="image" src="https://github.com/user-attachments/assets/7cb3c40d-ffe8-45fc-ab52-acd295fa067f" />

- [ ] In the gallery carousel (the one with the dissolve effect), shall we have mobile swipe scroll behavior as well? It seems to work on the other two but not this one

AliceR commented on app/content/themes/disasters.mdx:
```diff
@@ -7,6 +7,152 @@ image: "/images/themes/disasters/disasters-banner.png"
 <Section>
```
I still think it is weird that we use .mdx but write react/html - I added a talking point to the dev sync list about it!

AliceR commented on app/components/common/Section.tsx:
```diff
@@ -1,9 +1,15 @@
   return (
     <section className='usa-section'>
-      <div className='grid-container'>{children}</div>
+      <div className={`grid-container${full ? '-full' : ''}`}>{children}</div>
```
Is this a hacky way of just avoiding the grid for those cases? 
In that case I would encourage to just create a separate Section component, FullBleedSection or something along those lines :)

dzole0311 replied:
My intention was to make the `<Section />` span the full width of the page when the full prop is passed. The `grid-container-full` is a mistake, I’ve removed it. As you quoted from the USWDS docs: “If you would like the grid to span the full width of the page, do not use grid-container.” -> that’s what I was going for, just to omit the existing grid container class.

> In that case I would encourage to just create a separate Section component

I think it's nicer to have one adjustable `<Section />` rather then two separate components, which are almost the same.

AliceR reviewed (changes requested):
🚀 It is looking really good! Glad you found a cool library that supports both our transitions!

Only nitpicks and questions, no big concerns! Happy to approve after we had a chance to talk about the questions.

esthersuh commented:
@dzole0311 Looks great! For the second carousel, could we also create an extra carousel option where the carousel can move three cards at time (instead of one card at a time like what is shown here)? I think the card carousel where we can move one card at a time is still useful to have for future use cases, so please keep this one as well!

dzole0311 commented on app/components/common/Carousel/Carousel.tsx:
```diff
@@ -0,0 +1,140 @@
+              'padding-x-1',
+              !fade && visibleSlides.includes(index) && 'is-visible',
+              slideWidth === 'third' ? 'slide--third' : 'slide--full',
+            ]
```
Okay, tried my best to make it a bit less cryptic by improving some of the names.

dzole0311 commented on app/components/common/Carousel/Carousel.tsx:
```diff
@@ -0,0 +1,140 @@
+              .join(' ');
+
+            return (
+              <ul
```
The usa cards from uswds seem to be `<li />` elements, so I used `ul` to remove their list `dot` markers.

AliceR replied:
Huh, now I ran into the error: "`<ul>` and `<ol>` must only directly contain `<li>`, `<script>` or `<template>` elements" when trying to use the carousel with just images. We should rethink using `ul`, as copilot suggested...

AliceR replied:
Read up a bit deeper, and copilot was wrong: https://www.w3.org/WAI/tutorials/carousels/structure/
I'll wrap the images in `<li>`!

dzole0311 commented on app/components/common/Carousel/carousel.scss:
```diff
@@ -0,0 +1,37 @@
+@use 'uswds-core/src/styles/functions/units' as spacing;
+
+.carousel__viewport {
+  min-height: 470px;
```
I mainly used a min-height here to avoid a big layout shifts when hydrating the server rendered page with a client component. Now I reduced the height to come closer to the minimum height of the slider i saw in the designs (`400px`), but i can't find a fitting class from uswds. Maybe we can tweak the value again if need be.

dzole0311 commented on app/components/common/Carousel/carousel.scss:
```diff
@@ -0,0 +1,37 @@
+@use 'uswds-core/src/styles/functions/units' as spacing;
+
+.carousel__viewport {
```
USWDS util classes are preferred, but if it takes significant time to find a right one or if there's not one offered out of the box, i opt in for a custom class.

I cleaned up some of these now and used flex and justify-center from uswds for the flex layout.

dzole0311 commented:
@esthersuh I’ve added that mode now -> feel free to check the deployment page where I also added titles above each mode for better overview

> In the card carousel I think the steps are misaligned, it seems like there's an extra step in the logic that doesn't match the card position

@faustoperez and I chatted this morning and the ability to "move by 1 card" and "move by a group of 3 cards" modes satisfies the requirements for now.

AliceR commented:
Just reviewed on the go, changes looks all good! Only now on mobile I noticed a problem with the slides - I think we need to set a max width to the container and set overflow: hidden, otherwise the page width increases with every slide. Could you please double check the behavior? It might even be the same on desktop? (Try zooming out)

dzole0311 commented:
@AliceR Ah thanks for catching that, I think I fixed it now. Could you double check now before I remove the dummy/test content from the Disasters page?
